protocol-specs/generators/multisignature/index.js

```diff
@@ -0,0 +1,221 @@
+		return 0;
+	});
+
+const networkIdentifier =
+	'e48feb88db5b5cf5ad71d93cdcd1d879b6d5ed187a36b0002cc34e0ef9883255';
+
+const getId = transactionBytes => {
+	const transactionHash = hash(transactionBytes);
+	const bufferFromFirstEntriesReversed = getFirstEightBytesReversed(
+		transactionHash,
+	);
+	const transactionId = bufferToIntAsString(bufferFromFirstEntriesReversed);
+
+	return transactionId;
+};
+
+const assetToBytes = tx => {
+	const { mandatoryKeys, optionalKeys, numberOfSignatures } = tx.asset;
+	const mandatoryKeysBuffer = Buffer.from(mandatoryKeys.join(''), 'utf8');
+	const optionalKeysBuffer = Buffer.from(optionalKeys.join(''), 'utf8');
+	const assetBuffer = Buffer.concat([
+		intToBuffer(mandatoryKeys.length, 1),
+		mandatoryKeysBuffer,
+		intToBuffer(optionalKeys.length, 1),
+		optionalKeysBuffer,
+		intToBuffer(numberOfSignatures, 1),
+	]);
+
+	return assetBuffer;
+};
+
+const createSignatureObject = (txBuffer, account) => ({
+	// publicKey: account.publicKey,
+	signature: signData(
+		hash(Buffer.concat([hexToBuffer(networkIdentifier), txBuffer])),
+		account.passphrase,
+	),
+});
+
+const findAccountByKey = key =>
+	accounts.filter(account => account.publicKey === key);
+
+const generateValidMultisignatureRegistrationTransaction = () => {
+	// basic transaction
+	const tx = {
+		senderPublicKey: accounts[1].publicKey,
+		timestamp: 77045780,
+		type: 12,
+		asset: {
+			mandatoryKeys: [accounts[3].publicKey, accounts[2].publicKey],
+			optionalKeys: [accounts[4].publicKey, accounts[5].publicKey],
+			numberOfSignatures: 3,
+		},
+	};
+
+	sortKeysDescending(tx.asset.mandatoryKeys);
+	sortKeysDescending(tx.asset.optionalKeys);
+
+	const { signatures, ...registrationTx } = tx;
+	// Start building buffer
+	const transactionTimestamp = Buffer.alloc(4);
+	transactionTimestamp.writeIntBE(tx.timestamp, 0, 4);
+	let txBuffer = Buffer.concat([
+		Buffer.alloc(1, registrationTx.type),
+		transactionTimestamp,
+		hexToBuffer(tx.senderPublicKey),
+		intToBuffer(registrationTx.asset.numberOfSignatures, 1),
+		assetToBytes(registrationTx),
+	]);
+
+	registrationTx.signatures = [];
+	let index = 0;
+	// Mandatory keys sign
+	registrationTx.asset.mandatoryKeys.forEach(aKey => {
+		const account = findAccountByKey(aKey).pop();
+		registrationTx.signatures.push({
+			...createSignatureObject(txBuffer, account),
+			index,
+		});
+		index += 1;
+	});
+	// Optional keys sign
+	registrationTx.asset.optionalKeys.forEach(aKey => {
+		const account = findAccountByKey(aKey).pop();
+		registrationTx.signatures.push({
+			...createSignatureObject(txBuffer, account),
+			index,
+		});
+		index += 1;
+	});
+	// Serialize all signatures
+	registrationTx.signatures.forEach(aSignature => {
+		const signatureBuffer = Buffer.concat([
+			Buffer.alloc(1, aSignature.index),
+			hexToBuffer(aSignature.signature),
+		]);
+		txBuffer = Buffer.concat([txBuffer, signatureBuffer]);
+	});
+	// Sender signs whole transaction
+	const signature = signData(
```

Comment:
I think multi signature doesn't require signature

Reply:
This is for registration of multisig so in this case it does need it. It's not needed for transactions sent from a multisig

Reply:
ah that's true, if the owner account doesn't agree to change to multi sig, it sounds dangerous.